### PR TITLE
fix(core) `Checkpoint::drop` memory leak

### PR DESCRIPTION
**Pull request by oleonardolima**
based on #1997, depends on #2055

### Description

Fix memory leak bug in CheckPoint::drop by using Arc::into_inner.

Add tests (from the old PR) for memory leak + stack overflow when dropping CheckPoint.

### Notes to the reviewers

It should be merged after #2055.

### Changelog notice

```
### Fix:
- Fix memory leak bug in CheckPoint::drop by using `Arc::into_inner`.
```

#### All Submissions:

* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)

#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

### CODE REVIEW

File: .github/zizmor.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Zizmor config
+rules:
+  unpinned-uses:
+    config:
+      policies:
+        # Allow pin by ref/tag
+        actions-rust-lang/setup-rust-toolchain: ref-pin
```
Comment (ValuedMammal) [inline .github/zizmor.yml:7]: Since we're redefining the policy for `unpinned-uses`, we may need to reintroduce zizmor's default ref-pin policy for official GitHub actions.
```suggestion
        actions-rust-lang/setup-rust-toolchain: ref-pin
        actions/*: ref-pin
```

**Review by ValuedMammal — APPROVED** [review]
ACK a3bbba80066935ad976809d8df7ab5f8fefd0eff

File: .github/zizmor.yml (hunk 1 above)
Comment (oleonardolima) [reply .github/zizmor.yml:7]: Alright, I'll handle this on #2055.

**Review by evanlinjin — APPROVED** [review]
self-ACK 55161654ec107b7889ca3296aa64b32e3bc72eff

I rebased on the current state of #2055.

**Comment by oleonardolima** [comment]
I've rebased it on top of master, so it should be ready to go.

**Review by evanlinjin — APPROVED** [review]
ACK 0bd93dad114b90c3f08df9cc892625e250edb718
